features/dd-sdk-android-trace/src/test/kotlin/com/datadog/opentracing/SpanCleanerTest.kt

```diff
@@ -0,0 +1,204 @@
+    }
+
+    @Test
+    fun `M contain only one entry W add {same PendingTrace mutated, twice}`(forge: Forge) {
+        // Given
+        val pendingTrace = mockPendingTrace(forge)
+        testedCleaner.addPendingTrace(pendingTrace)
+        pendingTrace.addSpan(mock())
+        pendingTrace.addSpan(mock())
+        testedCleaner.addPendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testedCleaner.pendingTraces).hasSize(1)
+        assertThat(testedCleaner.pendingTraces.values).containsExactly(pendingTrace)
+
+        // When
+        testedCleaner.removePendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testedCleaner.pendingTraces).isEmpty()
+    }
+
+    @Test
+    fun `M contain 2 entries W add {2 PendingTrace instances with same root span}`(forge: Forge) {
+        // Given
+        val pendingTrace1 = mockPendingTrace(forge)
+        val pendingTrace2 = mockPendingTrace(forge)
+        val mock = mock<DDSpan>()
+        pendingTrace1.addSpan(mock)
+        pendingTrace2.addSpan(mock)
+        testedCleaner.addPendingTrace(pendingTrace1)
+        testedCleaner.addPendingTrace(pendingTrace2)
+
+        // Then
+        assertThat(testedCleaner.pendingTraces).hasSize(2)
+        assertThat(testedCleaner.pendingTraces.values).containsExactlyInAnyOrder(pendingTrace1, pendingTrace2)
+
+        // When
+        testedCleaner.removePendingTrace(pendingTrace1)
+        testedCleaner.removePendingTrace(pendingTrace2)
+
+        // Then
+        assertThat(testedCleaner.pendingTraces).isEmpty()
+    }
+
+    // endregion
+
+    // region concurrency
+
+    @Test
+    fun `M not leak any PendingTrace W add, mutate and remove { different threads }`(forge: Forge) {
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mockPendingTrace(forge) }
+        val addCountDownLatch = CountDownLatch(pendingTraces.size)
+        val removeCountDownLatch = CountDownLatch(pendingTraces.size)
+        val half = pendingTraces.size / 2
+        pendingTraces.take(half).forEach { pendingTrace ->
+            Thread {
+                testedCleaner.addPendingTrace(pendingTrace)
+                repeat(forge.anInt(min = 1, max = 10)) {
+                    pendingTrace.addSpan(mock())
+                }
+                addCountDownLatch.countDown()
+            }.start()
+        }
+        pendingTraces.subList(half, pendingTraces.size).forEach { pendingTrace ->
+            Thread {
+                if (forge.aBool()) {
+                    repeat(forge.anInt(min = 1, max = 10)) {
+                        pendingTrace.addSpan(mock())
+                    }
+                }
+                testedCleaner.addPendingTrace(pendingTrace)
+                repeat(forge.anInt(min = 1, max = 10)) {
+                    pendingTrace.addSpan(mock())
+                }
+                addCountDownLatch.countDown()
+            }.start()
+        }
+        addCountDownLatch.await(3, TimeUnit.SECONDS)
```

Comment:
`addCountDownLatch` will never reach zero, because we initialize it with `pendingTraces.size` counter, but `countDown` is called only `pendingTraces.size - half` times.